File: frontend/apps/ios/Pantopus/Features/Posts/PulsePostDetailViewModel.swift
//
//  PulsePostDetailViewModel.swift
//  Pantopus
//
//  Loads a single Pulse post, lets the user toggle reactions optimistic-
//  style, and posts new comments inline. Tap-targets for the multi-kind
//  reaction bar (Helpful/Heart/Going) all map to the single backend
//  `POST /:id/like` toggle today — see PR description for the discrepancy.
//

import Foundation
import Logging
import Observation

/// View-model state for the Pulse post detail screen.
public enum PulsePostDetailState: Sendable, Equatable {
    case loading
    case loaded(PulsePostDetailContent)
    case error(message: String)
}

/// Hydrated content for the Pulse post detail. Built in the VM from the
/// route handler's `PostDetailDTO` so the view sees only render-ready
/// values (formatted timestamps, mapped identity pillars, etc.).
public struct PulsePostDetailContent: Sendable, Equatable, Hashable {
    public let post: PostDetailDTO
    public let authorDisplayName: String
    public let authorAvatarURL: URL?
    public let authorIdentity: IdentityPillar
    public let authorVerified: Bool
    public let timeAndLocality: String
    public let intent: PostIntent
    public let mediaURLs: [URL]
    public let reactions: PostReactionCounts
    public let comments: [PostCommentRow]
    public let hiddenReplyCount: Int

    public init(
        post: PostDetailDTO,
        authorDisplayName: String,
        authorAvatarURL: URL?,
        authorIdentity: IdentityPillar,
        authorVerified: Bool,
        timeAndLocality: String,
        intent: PostIntent,
        mediaURLs: [URL],
        reactions: PostReactionCounts,
        comments: [PostCommentRow],
        hiddenReplyCount: Int
    ) {
        self.post = post
        self.authorDisplayName = authorDisplayName
        self.authorAvatarURL = authorAvatarURL
        self.authorIdentity = authorIdentity
        self.authorVerified = authorVerified
        self.timeAndLocality = timeAndLocality
        self.intent = intent
        self.mediaURLs = mediaURLs
        self.reactions = reactions
        self.comments = comments
        self.hiddenReplyCount = hiddenReplyCount
    }
}

/// Loads + mutates a single Pulse post.
@MainActor
@Observable
public final class PulsePostDetailViewModel {
    /// Render-state.
    public private(set) var state: PulsePostDetailState = .loading

    /// Inline composer's draft text. Bound from the view.
    public var composerText: String = ""

    /// True while a comment is in flight.
    public private(set) var isSendingComment: Bool = false

    /// Transient error string surfaced in the toast overlay.
    public var toastMessage: String?

    /// True when the entire reply thread is expanded.
    public private(set) var showingAllReplies: Bool = false

    private let postId: String
    private let client: APIClient
    private let logger = Logger(label: "app.pantopus.ios.PulsePostDetail")
    private let maxInitialReplies = 3

    init(postId: String, client: APIClient = .shared) {
        self.postId = postId
        self.client = client
    }

    /// First-load entry. Re-run via `refresh()` after a pull-to-refresh.
    public func load() async {
        state = .loading
        await fetch()
    }

    /// Pull-to-refresh.
    public func refresh() async {
        await fetch()
    }

    /// Expand the truncated reply list.
    public func showMoreReplies() {
        guard case let .loaded(content) = state else { return }
        showingAllReplies = true
        state = .loaded(rebuildContent(from: content.post))
    }

    /// Tap one of the reaction pills. Only `.helpful` is wired to a
    /// backend route today; the `ReactionsBar` view renders the other
    /// kinds as display-only so this method only ever sees `.helpful`,
    /// but we keep the guard as a safety net.
    public func tapReaction(_ kind: PostReactionKind) async {
        guard case var .loaded(content) = state else { return }
        guard kind.isBackendWired else { return }

        // Optimistic update on the helpful (likes) bucket.
        var optimistic = content.reactions
        let wasOn = optimistic.userReaction == .helpful
        if wasOn {
            optimistic.helpful = max(0, optimistic.helpful - 1)
            optimistic.userReaction = nil
        } else {
            optimistic.helpful += 1
            optimistic.userReaction = .helpful
        }
        content = PulsePostDetailContent(
            post: content.post,
            authorDisplayName: content.authorDisplayName,
            authorAvatarURL: content.authorAvatarURL,
            authorIdentity: content.authorIdentity,
            authorVerified: content.authorVerified,
            timeAndLocality: content.timeAndLocality,
            intent: content.intent,
            mediaURLs: content.mediaURLs,
            reactions: optimistic,
            comments: content.comments,
            hiddenReplyCount: content.hiddenReplyCount
        )
        state = .loaded(content)

        do {
            let response = try await client.request(
                PostsEndpoints.toggleLike(id: postId),
                as: PostLikeResponse.self
            )
            // Reconcile with server truth.
            var reconciled = optimistic
            reconciled.helpful = response.likeCount
            reconciled.userReaction = response.liked ? .helpful : nil
            content = PulsePostDetailContent(
                post: content.post,
                authorDisplayName: content.authorDisplayName,
                authorAvatarURL: content.authorAvatarURL,
                authorIdentity: content.authorIdentity,
                authorVerified: content.authorVerified,
                timeAndLocality: content.timeAndLocality,
                intent: content.intent,
                mediaURLs: content.mediaURLs,
                reactions: reconciled,
                comments: content.comments,
                hiddenReplyCount: content.hiddenReplyCount
            )
            state = .loaded(content)
        } catch {
            logger.warning("Reaction toggle failed: \(error)")
            toastMessage = "Couldn't update your reaction"
            // Rollback.
            var rolled = optimistic
            if wasOn {
                rolled.helpful += 1
                rolled.userReaction = .helpful
            } else {
                rolled.helpful = max(0, rolled.helpful - 1)
                rolled.userReaction = nil
            }
            content = PulsePostDetailContent(
                post: content.post,
                authorDisplayName: content.authorDisplayName,
                authorAvatarURL: content.authorAvatarURL,
                authorIdentity: content.authorIdentity,
                authorVerified: content.authorVerified,
                timeAndLocality: content.timeAndLocality,
                intent: content.intent,
                mediaURLs: content.mediaURLs,
                reactions: rolled,
                comments: content.comments,
                hiddenReplyCount: content.hiddenReplyCount
            )
            state = .loaded(content)
        }
    }

    /// Submit the composer's current text as a new top-level comment.
    /// On success we re-fetch the post so the comment list, count, and
    /// any backend-side mutations (rate-limit windows, derived fields)
    /// stay in sync.
    public func sendComment() async {
        let body = composerText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !body.isEmpty, case .loaded = state, !isSendingComment else { return }
        isSendingComment = true
        defer { isSendingComment = false }
        let req = PostCommentRequest(comment: body, parentCommentId: nil)
        do {
            _ = try await client.request(
                PostsEndpoints.createComment(id: postId, body: req),
                as: PostCommentCreateResponse.self
            )
            composerText = ""
            await fetch()
        } catch {
            logger.warning("Comment send failed: \(error)")
            toastMessage = "Couldn't post your comment"
        }
    }

    // MARK: - Internal helpers

    private func fetch() async {
        do {
            let response = try await client.request(
                PostsEndpoints.detail(id: postId),
                as: PostDetailResponse.self
            )
            state = .loaded(rebuildContent(from: response.post))
        } catch let error as APIError {
            logger.warning("Post detail load failed: \(error)")
            state = .error(message: friendlyMessage(for: error))
        } catch {
            logger.warning("Post detail load failed: \(error)")
            state = .error(message: "Something went wrong")
        }
    }

    private func rebuildContent(from post: PostDetailDTO) -> PulsePostDetailContent {
        let displayName = post.creator?.displayName ?? "Pantopus user"
        let avatarURL = post.creator?.profilePictureURL.flatMap(URL.init(string:))
        let mediaURLs = post.mediaURLs.compactMap(URL.init(string:))
        let identity: IdentityPillar = mapAccountType(post.creator?.accountType)
        let timeAndLocality = formatTimeAndLocality(
            createdAt: post.createdAt,
            locality: post.creator?.locality ?? post.home?.city
        )
        let intent = PostIntent.from(purpose: post.purpose, postType: post.postType)
        let reactions = PostReactionCounts(
            helpful: post.likeCount,
            heart: 0,
            going: 0,
            userReaction: post.userHasLiked ? .helpful : nil
        )
        let (rows, hidden) = buildCommentRows(post.comments)
        return PulsePostDetailContent(
            post: post,
            authorDisplayName: displayName,
            authorAvatarURL: avatarURL,
            authorIdentity: identity,
            // TODO(backend): backend `CREATOR_SELECT` for posts does not
            // include the `verified` column today, so the post header
            // can never show a verified badge. Wire this up once
            // feedService.js#CREATOR_SELECT joins `verified`.
            authorVerified: false,
            timeAndLocality: timeAndLocality,
            intent: intent,
            mediaURLs: mediaURLs,
            reactions: reactions,
            comments: rows,
            hiddenReplyCount: hidden
        )
    }

    private func mapAccountType(_ accountType: String?) -> IdentityPillar {
        switch accountType {
        case "business": .business
        case "home": .home
        default: .personal
        }
    }

    private func buildCommentRows(_ comments: [PostCommentDTO]) -> (rows: [PostCommentRow], hidden: Int) {
        // Group by parent → flatten with indentLevel capped at 1.
        let topLevel = comments.filter { $0.parentCommentId == nil && !$0.isDeleted }
        let repliesByParent = Dictionary(
            grouping: comments.filter { $0.parentCommentId != nil && !$0.isDeleted }
        ) { $0.parentCommentId ?? "" }

        var rows: [PostCommentRow] = []
        var visibleReplyCount = 0
        var totalReplyCount = 0

        for parent in topLevel {
            rows.append(row(from: parent, indent: 0))
            let replies = (repliesByParent[parent.id] ?? [])
                .sorted { ($0.createdAt) < ($1.createdAt) }
            totalReplyCount += replies.count
            let cap = showingAllReplies ? replies.count : min(maxInitialReplies, replies.count)
            for reply in replies.prefix(cap) {
                rows.append(row(from: reply, indent: 1))
                visibleReplyCount += 1
            }
        }

        let hidden = showingAllReplies ? 0 : max(0, totalReplyCount - visibleReplyCount)
        return (rows, hidden)
    }

    private func row(from comment: PostCommentDTO, indent: Int) -> PostCommentRow {
        let name = comment.author?.displayName ?? "Pantopus user"
        return PostCommentRow(
            id: comment.id,
            authorName: name,
            authorAvatarURL: comment.author?.profilePictureURL.flatMap(URL.init(string:)),
            authorIdentity: mapAccountType(comment.author?.accountType),
            body: comment.comment,
            timestamp: relativeTimestamp(comment.createdAt),
            indentLevel: indent,
            authorUserId: comment.author?.id
        )
    }

    private func formatTimeAndLocality(createdAt: String, locality: String?) -> String {
        let ts = relativeTimestamp(createdAt)
        if let locality, !locality.isEmpty {
            return "\(ts) · \(locality)"
        }
        return ts
    }

    private func relativeTimestamp(_ iso: String) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let date = formatter.date(from: iso) ?? ISO8601DateFormatter().date(from: iso) ?? Date()
        let elapsed = Date().timeIntervalSince(date)
        switch elapsed {
        case ..<60: return "Just now"
        case ..<3600: return "\(Int(elapsed / 60))m ago"
        case ..<86400: return "\(Int(elapsed / 3600))h ago"
        case ..<604_800: return "\(Int(elapsed / 86400))d ago"
        default:
            let display = DateFormatter()
            display.dateStyle = .medium
            display.timeStyle = .none
            return display.string(from: date)
        }
    }

    private func friendlyMessage(for error: APIError) -> String {
        switch error {
        case .notFound: "We couldn't find this post."
        case .forbidden: "You don't have access to this post."
        case .transport: "Check your connection and try again."
        default: "Something went wrong. Try again."
        }
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/posts/PulsePostDetailViewModel.kt
@file:Suppress("MagicNumber", "LongMethod", "PackageNaming", "TooGenericExceptionCaught", "TooManyFunctions")

package app.pantopus.android.ui.screens.posts

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.posts.PostCommentDto
import app.pantopus.android.data.api.models.posts.PostCommentRequest
import app.pantopus.android.data.api.models.posts.PostDetailDto
import app.pantopus.android.data.api.models.posts.PostReactionKind
import app.pantopus.android.data.api.net.NetworkError
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.auth.AuthRepository
import app.pantopus.android.data.posts.PostsRepository
import app.pantopus.android.ui.components.IdentityPillar
import app.pantopus.android.ui.screens.shared.content_detail.bodies.PostCommentRow
import app.pantopus.android.ui.screens.shared.content_detail.bodies.PostReactionCounts
import app.pantopus.android.ui.screens.shared.content_detail.headers.PostIntent
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.Duration
import java.time.Instant
import java.time.format.DateTimeFormatter
import javax.inject.Inject

/** Nav-arg key for the Pulse post ID. */
const val PULSE_POST_DETAIL_ID_KEY = "postId"

/** Render-ready payload for the Pulse post detail screen. */
data class PulsePostDetailContent(
    val post: PostDetailDto,
    val authorDisplayName: String,
    val authorAvatarUrl: String?,
    val authorIdentity: IdentityPillar,
    val authorVerified: Boolean,
    val timeAndLocality: String,
    val intent: PostIntent,
    val mediaUrls: List<String>,
    val reactions: PostReactionCounts,
    val comments: List<PostCommentRow>,
    val hiddenReplyCount: Int,
)

/** Observed UI state for the Pulse post detail screen. */
sealed interface PulsePostDetailUiState {
    data object Loading : PulsePostDetailUiState

    data class Loaded(val content: PulsePostDetailContent) : PulsePostDetailUiState

    data class Error(val message: String) : PulsePostDetailUiState
}

/** Loads + mutates a single Pulse post. */
@HiltViewModel
class PulsePostDetailViewModel
    @Inject
    constructor(
        private val repo: PostsRepository,
        private val authRepo: AuthRepository,
        savedStateHandle: SavedStateHandle,
    ) : ViewModel() {
        private val postId: String =
            requireNotNull(savedStateHandle[PULSE_POST_DETAIL_ID_KEY]) {
                "PulsePostDetailViewModel requires a '$PULSE_POST_DETAIL_ID_KEY' nav arg."
            }

        private val _state = MutableStateFlow<PulsePostDetailUiState>(PulsePostDetailUiState.Loading)
        val state: StateFlow<PulsePostDetailUiState> = _state.asStateFlow()

        private val _composerText = MutableStateFlow("")
        val composerText: StateFlow<String> = _composerText.asStateFlow()

        private val _isSendingComment = MutableStateFlow(false)
        val isSendingComment: StateFlow<Boolean> = _isSendingComment.asStateFlow()

        private val _toastMessage = MutableStateFlow<String?>(null)
        val toastMessage: StateFlow<String?> = _toastMessage.asStateFlow()

        private val _showsOverflowMenu = MutableStateFlow(false)

        /**
         * Bound to the screen's overflow modal so the Edit action sheet
         * pops when the owner taps the top-bar's more-horizontal icon.
         */
        val showsOverflowMenu: StateFlow<Boolean> = _showsOverflowMenu.asStateFlow()

        private val showingAllReplies = MutableStateFlow(false)

        private val maxInitialReplies = 3

        /**
         * True when the signed-in user authored the post on screen. The
         * screen uses this to gate the Edit overflow action.
         */
        val isOwner: Boolean
            get() {
                val loaded = _state.value as? PulsePostDetailUiState.Loaded ?: return false
                val signedIn = authRepo.state.value as? AuthRepository.State.SignedIn ?: return false
                return loaded.content.post.userId == signedIn.user.id
            }

        fun openOverflowMenu() {
            _showsOverflowMenu.value = true
        }

        fun dismissOverflowMenu() {
            _showsOverflowMenu.value = false
        }

        /** First-load entry. */
        fun load() {
            if (_state.value is PulsePostDetailUiState.Loaded) return
            refresh()
        }

        /** Pull-to-refresh / retry. */
        fun refresh() {
            _state.value = PulsePostDetailUiState.Loading
            viewModelScope.launch { fetch() }
        }

        fun setComposerText(text: String) {
            _composerText.value = text
        }

        fun dismissToast() {
            _toastMessage.value = null
        }

        fun showMoreReplies() {
            val loaded = _state.value as? PulsePostDetailUiState.Loaded ?: return
            showingAllReplies.value = true
            _state.value = PulsePostDetailUiState.Loaded(rebuildContent(loaded.content.post))
        }

        /**
         * Tap one of the reaction pills. Only `.Helpful` is wired to a
         * backend route today; the `ReactionsBar` view renders the
         * other kinds as display-only so this method only ever sees
         * `.Helpful`, but we keep the guard as a safety net.
         */
        fun tapReaction(kind: PostReactionKind) {
            val loaded = _state.value as? PulsePostDetailUiState.Loaded ?: return
            if (!kind.isBackendWired) return
            val initialReactions = loaded.content.reactions
            val wasOn = initialReactions.userReaction == PostReactionKind.Helpful
            val optimistic =
                initialReactions.copy(
                    helpful = if (wasOn) (initialReactions.helpful - 1).coerceAtLeast(0) else initialReactions.helpful + 1,
                    userReaction = if (wasOn) null else PostReactionKind.Helpful,
                )
            _state.value = PulsePostDetailUiState.Loaded(loaded.content.copy(reactions = optimistic))

            viewModelScope.launch {
                when (val result = repo.toggleLike(postId)) {
                    is NetworkResult.Success -> {
                        val reconciled =
                            optimistic.copy(
                                helpful = result.data.likeCount,
                                userReaction = if (result.data.liked) PostReactionKind.Helpful else null,
                            )
                        _state.update { current ->
                            (current as? PulsePostDetailUiState.Loaded)?.let {
                                PulsePostDetailUiState.Loaded(it.content.copy(reactions = reconciled))
                            } ?: current
                        }
                    }
                    is NetworkResult.Failure -> {
                        _toastMessage.value = "Couldn't update your reaction"
                        _state.update { current ->
                            (current as? PulsePostDetailUiState.Loaded)?.let {
                                PulsePostDetailUiState.Loaded(
                                    it.content.copy(reactions = initialReactions),
                                )
                            } ?: current
                        }
                    }
                }
            }
        }

        /** Submit the composer's current text as a new top-level comment. */
        fun sendComment() {
            val body = _composerText.value.trim()
            if (body.isEmpty() || _isSendingComment.value) return
            _isSendingComment.value = true
            viewModelScope.launch {
                val req = PostCommentRequest(comment = body, parentCommentId = null)
                when (repo.createComment(postId, req)) {
                    is NetworkResult.Success -> {
                        _composerText.value = ""
                        fetch()
                    }
                    is NetworkResult.Failure -> {
                        _toastMessage.value = "Couldn't post your comment"
                    }
                }
                _isSendingComment.value = false
            }
        }

        // MARK: - Internal helpers

        private suspend fun fetch() {
            when (val result = repo.detail(postId)) {
                is NetworkResult.Success -> {
                    _state.value = PulsePostDetailUiState.Loaded(rebuildContent(result.data.post))
                }
                is NetworkResult.Failure -> {
                    _state.value = PulsePostDetailUiState.Error(friendlyMessage(result.error))
                }
            }
        }

        private fun rebuildContent(post: PostDetailDto): PulsePostDetailContent {
            val (rows, hidden) = buildCommentRows(post.comments)
            val identity = mapAccountType(post.creator?.accountType)
            return PulsePostDetailContent(
                post = post,
                authorDisplayName = post.creator?.displayName ?: "Pantopus user",
                authorAvatarUrl = post.creator?.profilePictureUrl,
                authorIdentity = identity,
                // TODO(backend): backend `CREATOR_SELECT` for posts does
                // not include the `verified` column today, so the post
                // header can never show a verified badge. Wire this up
                // once feedService.js#CREATOR_SELECT joins `verified`.
                authorVerified = false,
                timeAndLocality =
                    formatTimeAndLocality(
                        createdAt = post.createdAt,
                        locality = post.creator?.locality ?: post.home?.city,
                    ),
                intent = PostIntent.from(post.purpose, post.postType),
                mediaUrls = post.mediaUrls,
                reactions =
                    PostReactionCounts(
                        helpful = post.likeCount,
                        heart = 0,
                        going = 0,
                        userReaction = if (post.userHasLiked) PostReactionKind.Helpful else null,
                    ),
                comments = rows,
                hiddenReplyCount = hidden,
            )
        }

        private fun mapAccountType(type: String?): IdentityPillar =
            when (type) {
                "business" -> IdentityPillar.Business
                "home" -> IdentityPillar.Home
                else -> IdentityPillar.Personal
            }

        private fun buildCommentRows(comments: List<PostCommentDto>): Pair<List<PostCommentRow>, Int> {
            val visible = comments.filter { !it.isDeleted }
            val topLevel = visible.filter { it.parentCommentId == null }
            val repliesByParent =
                visible.filter { it.parentCommentId != null }
                    .groupBy { it.parentCommentId ?: "" }
                    .mapValues { (_, list) -> list.sortedBy { it.createdAt } }

            val rows = mutableListOf<PostCommentRow>()
            var visibleReplyCount = 0
            var totalReplyCount = 0

            for (parent in topLevel) {
                rows += rowFrom(parent, indent = 0)
                val replies = repliesByParent[parent.id].orEmpty()
                totalReplyCount += replies.size
                val cap =
                    if (showingAllReplies.value) {
                        replies.size
                    } else {
                        minOf(maxInitialReplies, replies.size)
                    }
                for (reply in replies.take(cap)) {
                    rows += rowFrom(reply, indent = 1)
                    visibleReplyCount++
                }
            }
            val hidden =
                if (showingAllReplies.value) {
                    0
                } else {
                    (totalReplyCount - visibleReplyCount).coerceAtLeast(0)
                }
            return rows to hidden
        }

        private fun rowFrom(
            comment: PostCommentDto,
            indent: Int,
        ): PostCommentRow =
            PostCommentRow(
                id = comment.id,
                authorName = comment.author?.displayName ?: "Pantopus user",
                authorAvatarUrl = comment.author?.profilePictureUrl,
                authorIdentity = mapAccountType(comment.author?.accountType),
                body = comment.comment,
                timestamp = relativeTimestamp(comment.createdAt),
                reactionCount = comment.likeCount ?: 0,
                userReacted = comment.userHasLiked ?: false,
                indentLevel = indent,
                authorUserId = comment.author?.id,
            )

        private fun formatTimeAndLocality(
            createdAt: String,
            locality: String?,
        ): String {
            val ts = relativeTimestamp(createdAt)
            return if (!locality.isNullOrEmpty()) "$ts · $locality" else ts
        }

        private fun relativeTimestamp(iso: String?): String {
            if (iso.isNullOrEmpty()) return ""
            val instant =
                try {
                    Instant.parse(iso)
                } catch (_: Throwable) {
                    return ""
                }
            val elapsed = Duration.between(instant, Instant.now())
            val seconds = elapsed.seconds
            return when {
                seconds < 60 -> "Just now"
                seconds < 3_600 -> "${seconds / 60}m ago"
                seconds < 86_400 -> "${seconds / 3_600}h ago"
                seconds < 604_800 -> "${seconds / 86_400}d ago"
                else ->
                    DateTimeFormatter.ISO_LOCAL_DATE.format(
                        instant.atZone(java.time.ZoneId.systemDefault()).toLocalDate(),
                    )
            }
        }

        private fun friendlyMessage(error: NetworkError): String =
            when (error) {
                NetworkError.NotFound -> "We couldn't find this post."
                NetworkError.Forbidden -> "You don't have access to this post."
                is NetworkError.Transport -> "Check your connection and try again."
                else -> "Something went wrong. Try again."
            }
    }
